source/common/request_info/request_info_impl.h

```diff
@@ -15,39 +17,80 @@ struct RequestInfoImpl : public RequestInfo {
 
   RequestInfoImpl(Http::Protocol protocol) : RequestInfoImpl() { protocol_ = protocol; }
 
-  // RequestInfo::RequestInfo
-  SystemTime startTime() const override { return start_time_; }
+  const SystemTime& startTime() const override { return start_time_; }
+
+  const MonotonicTime& startTimeMonotonic() const override { return start_time_monotonic_; }
+
+  const Optional<MonotonicTime>& lastDownstreamRxByteReceived() const override {
+    return last_rx_byte_received_;
+  }
+
+  void lastDownstreamRxByteReceived(MonotonicTime time) override {
+    last_rx_byte_received_.value(time);
+  }
+
+  const Optional<MonotonicTime>& firstUpstreamTxByteSent() const override {
+    return first_upstream_tx_byte_sent_;
+  }
+
+  void firstUpstreamTxByteSent(MonotonicTime time) override {
+    first_upstream_tx_byte_sent_.value(time);
+  }
+
+  const Optional<MonotonicTime>& lastUpstreamTxByteSent() const override {
+    return last_upstream_tx_byte_sent_;
+  }
+
+  void lastUpstreamTxByteSent(MonotonicTime time) override {
+    last_upstream_tx_byte_sent_.value(time);
+  }
+
+  const Optional<MonotonicTime>& firstUpstreamRxByteReceived() const override {
+    return first_upstream_rx_byte_received_;
+  }
+
+  void firstUpstreamRxByteReceived(MonotonicTime time) override {
+    first_upstream_rx_byte_received_.value(time);
+  }
 
-  const Optional<std::chrono::microseconds>& requestReceivedDuration() const override {
-    return request_received_duration_;
+  const Optional<MonotonicTime>& lastUpstreamRxByteReceived() const override {
+    return last_upstream_rx_byte_received_;
   }
-  void requestReceivedDuration(MonotonicTime time) override {
-    request_received_duration_ =
-        std::chrono::duration_cast<std::chrono::microseconds>(time - start_time_monotonic_);
+
+  void lastUpstreamRxByteReceived(MonotonicTime time) override {
+    last_upstream_rx_byte_received_.value(time);
   }
 
-  const Optional<std::chrono::microseconds>& responseReceivedDuration() const override {
-    return response_received_duration_;
+  const Optional<MonotonicTime>& firstDownstreamTxByteSent() const override {
+    return first_downstream_tx_byte_sent_;
   }
-  void responseReceivedDuration(MonotonicTime time) override {
-    response_received_duration_ =
-        std::chrono::duration_cast<std::chrono::microseconds>(time - start_time_monotonic_);
+
+  void firstDownstreamTxByteSent(MonotonicTime time) override {
+    first_downstream_tx_byte_sent_.value(time);
+  }
+
+  const Optional<MonotonicTime>& lastDownstreamTxByteSent() const override {
+    return last_downstream_tx_byte_sent_;
   }
 
+  void lastDownstreamTxByteSent(MonotonicTime time) override {
+    last_downstream_tx_byte_sent_.value(time);
+  }
+
+  void finalize(MonotonicTime time) override { end_time_.value(time); }
```

Comment:
Should this be renamed to finalTimeMonotonic (and in interface, of course) to match the pattern in the rest of the interface?

Reply:
Was leaving it generic if there's other finalization tasks to be added in the future. YAGNI, I suppose; will rename.